src/Runner/Parallel/ParallelConfig.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Runner\Parallel;
+
+use Fidry\CpuCoreCounter\CpuCoreCounter;
+use Fidry\CpuCoreCounter\Finder\DummyCpuCoreFinder;
+use Fidry\CpuCoreCounter\Finder\FinderRegistry;
+
+/**
+ * @author Greg Korba <greg@codito.dev>
+ */
+final class ParallelConfig
+{
+    private const DEFAULT_FILES_PER_PROCESS = 10;
```

Comment:
I wonder about this. We have value that will fit one size of repo, and depends on repo size user may need to customise it manually to best fit. I wonder if we can automate that.

If we have 8 maxProcessess and 8_000 files to be processed, it may be beneficial to have 1_000 files per process, or being ready for non-equal effort per file, and set it to sth smaller than 1_000, but still bigger than 10.
Eg sth like `min(100, max(10, $filesCount / $cpuCount / 4))`

Reply:
And where's the problem? You just use `ParallelConfig::detect(1000)` and don't use default value for the chunk size. But anyway, it does not make sense to use such big chunks, because it would make progress output jump between values (because progress events are dispatched in the main process after receiving response from the worker). It would be possible to report file analysis result for each file separately rather than sending report for whole chunk, but again it would require making `Runner::processFile()` public so we can collect result data after each iteration. Anyway, I think it's not a good idea because it would mean much bigger data transferred between parallelisation operator and the worker in single request (especially when it comes to analysis report sent from worker).

It also does not make sense IMHO to use different size of chunks for workers. I don't see the actual value in it.

Reply:
> And where's the problem? 

bootstrapping the CLI app is having the cost on it's own. if it would not, you would always bootstrap it for each file independently.

as the end goal I would love to not have any config for the feature needed to be manually provided, and autodiscovery of config makes the work best for given case - including the pool size.

I see the reason with worker reports all at once - would it be possible for it to report progress partially step by step?
extra: is worker a separated process or thread ? if later, we can reuse existing events bug to just main thread about progress.

Reply:
> as the end goal I would love to not have any config for the feature needed to be manually provided, and autodiscovery of config makes the work best for given case - including the pool size.

It's only the matter of [fallback value in config resolver](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7777/files#diff-c0526b685281b1ce96efcaa2e77ae9658fc6c59b4b40975f5cde342f1678f98dR285). For now I wanted to use sequential analysis by default and provide parallel runner as an opt-in. After verifying the usage in real world apps we can switch to "parallel with auto-detection by default".

> I see the reason with worker reports all at once - would it be possible for it to report progress partially step by step?

As I said, it's possible but it requires further changes in the `Runner`, so iteration is done on worker's side and result can be reported after each step. But I don't see much gain in it, because IMHO there's no point in sending large file chunks - worker's process has timer which is resetting on each request (file chunk), so requesting big file chunks would require bumping process timeout too. It's much safer to use smaller chunks with smaller timeout and just iterate on runner's side.

> extra: is worker a separated process or thread ? if later, we can reuse existing events bug to just main thread about progress.

It's a process, so the data is separated. Communication goes through TCP socket AFAIR (main CLI command initialises server, spawns worker processes, they connect to server and create `Runner` in sequential mode, analysing what's requested by the operator). So events are caught on the worker's side, they are sent via React connection and then are re-created on operator's side and dispatched in the main process, to ensure same data flow as in sequential analysis.

Reply:
:+1: for points 1 and 3.

for 2 (and bit 3) - if we have communication via TCP socket, is it really a potential bottleneck (that we would need to extend timeouts) to send multiple small updates instead of all-in-one update?
(only asking, aiming to understand proposed solution)

Reply:
Small chunks have IMHO several advantages. Besides the ones already pointed, there is also better job distribution: when you have huge project with many files, it's better to process them in smaller chunks so multiple workers consume it until the end rather than passing huge chunks which may lead to situation where few workers would get small files and would end work (no more files in the main queue) and one worker would receive 1000 huge files and would process them alone. Smaller chunks would not end up like this, because more workers would consume these large files and analysis would be faster. Also, with large chunks it's really hard to determine timeout value, because you never know how long would take the biggest chunk (in terms of sum of files' size).

Reply:
That's why I was wondering not about using 1000, but 250 - still having chance to fairly distribute problematic files and not have one handing job while everyone else finished, but still avoiding the overhead of initialising full app for 10 files only.

ah, so the timeout is not to pass any msg from worker to supervisor, but for whole worker to finish the job? if we would have worker reporting each file progress, would it prevent having big timeouts? (ie it's timeout till we hear anything back from worker, or timeout to worker to finish everything?)

[not judging for next steps direction, simply in need to understand how this work]

Reply:
It's a time between operator's request to worker and worker's response, so in current approach it's a "chunk timeout". So if we changed the way how worker reports analysis to "per file", timeout could be small.

Worker is bootstrapped only once, then starts the event loop and just listens for requests. So it actually does not matter what chunk size you choose - the only difference would be the amount of TCP requests/responses. For example: you have 8000 files and you want to spawn 8 workers with 1000-files chunks, there will be only one request-response per chunk, and 1 app bootstrap too. On the other hand, you could use chunk size 50 and just make multiple request-response per worker and achieve the same thing. And when you have 250 files in repo it does not make sense to use 250-files chunk size because it would end like sequential analysis with one worker... Really, I think there's no point in discussing it, since the `ParallelConfig` allows you to explicitly set the chunk size and it does not matter what's the default chunk size 🙂. Also `ParallelConfig::detect()` supports optional parameters so you can detect CPU cores but still use specified chunk size or timeout.

Reply:
> Really, I think there's no point in discussing it

huge point it is in discussing. I want to understand solution before merging it.

With 10 files chunk, would we spawn 8 workers and reuse them, meaning we instantiate `Runner` 8 times and simply call for next `->fix()` multiple times, or we would spawn fresh worker again and again, after each chunk of 10 files?
if we would instantiate only 8 workers ever (and not limiting up to 8 parallel workers, but let them die and we recreate them), it is solving all my concerns.

Reply:
Yes, workers are spawned only once and only `setFileIterator()` and `fix()` is called on existing Runner instance for each chunk (and then the result is collected and returned to the operator). It is rather clear when you look at `WorkerCommand` and my previous comments 😅.